utils/fdb_wrapper_cpp/src/fdbcs.cpp

```diff
@@ -206,6 +223,312 @@ std::shared_ptr<FDBDataBase> DataBaseCreator::createDataBase(const std::string c
+    {
+      const auto& currentKey = currentKeys[i];
+      auto& block = map[currentKey];
+      for (uint32_t j = 0; j < numKeysInBlock_ && nextKeysIt < nextLevelKeyNum; ++j, ++nextKeysIt)
+      {
+        const auto& nextKey = nextLevelKeys[nextKeysIt];
+        insertKey(block, nextKey);
+        map[nextKey] = {0, std::string()};
+      }
+      if (currentTnxSize + (keySizeInBytes_ + block.second.size()) >= maxTnxSize_)
+      {
+        RETURN_ON_ERROR(commitKeys(dataBase, map, keysInTnx));
+        currentTnxSize = 0;
+        keysInTnx.clear();
+      }
+      currentTnxSize += block.second.size() + keySizeInBytes_;
+      keysInTnx.push_back(currentKey);
+    }
+    currentKeys = std::move(nextLevelKeys);
+  }
+
+  size_t offset = 0;
+  for (uint32_t i = 0; i < numDataBlocks; ++i)
+  {
+    const auto& currentKey = currentKeys[i];
+    auto& block = map[currentKey];
+    offset = insertData(block, blob, offset);
+    if (currentTnxSize + (keySizeInBytes_ + block.second.size()) >= maxTnxSize_)
+    {
+      RETURN_ON_ERROR(commitKeys(dataBase, map, keysInTnx));
+      currentTnxSize = 0;
+      keysInTnx.clear();
+    }
+    keysInTnx.push_back(currentKey);
+    currentTnxSize += block.second.size() + keySizeInBytes_;
+  }
+
+  if (currentTnxSize)
+    RETURN_ON_ERROR(commitKeys(dataBase, map, keysInTnx));
+
+  return true;
+}
+
+std::pair<bool, Keys> BlobHandler::getKeysFromBlock(const Block& block)
+{
+  Keys keys;
+  const auto& blockData = block.second;
+  if (blockData.size() > blockSizeInBytes_)
+    return {false, {""}};
+
+  uint32_t offset = 1;
+  for (uint32_t i = 0; i < numKeysInBlock_ && offset + keySizeInBytes_ <= blockData.size(); ++i)
+  {
+    Key key(blockData.begin() + offset, blockData.begin() + offset + keySizeInBytes_);
+    keys.push_back(std::move(key));
+    offset += keySizeInBytes_;
+  }
+
+  return {true, keys};
+}
+
+bool BlobHandler::isDataBlock(const Block& block)
+{
+  return block.second.compare(0, keyBlockIdentifier.size(), keyBlockIdentifier) != 0;
+}
+
+std::pair<bool, std::string> BlobHandler::readBlob(std::shared_ptr<FDBCS::FDBDataBase> database,
+                                                   const ByteArray& key)
+{
+  Keys currentKeys{key};
+  bool dataBlockReached = false;
+  int level = 0;
+
+  while (!dataBlockReached)
+  {
+    auto tnx = database->createTransaction();
+    if (!tnx)
+      return {false, ""};
+
+    std::vector<Block> blocks;
+    for (const auto& key : currentKeys)
+    {
+      auto p = tnx->get(key);
+      if (!p.first)
+        return {false, ""};
+
+      Block block{0, p.second};
+      if (isDataBlock(block))
+      {
+        dataBlockReached = true;
+        break;
+      }
+      blocks.push_back(block);
+    }
+
+    if (dataBlockReached)
+      break;
+
+    Keys nextKeys;
+    for (const auto& block : blocks)
+    {
+      auto keysPair = getKeysFromBlock(block);
+      if (!keysPair.first)
```

Comment:
Why don't you check for `isDataBlock` is true iterating blocks on this level of a tree?
For example you do check this in the previous loop over `currentKeys`.

Reply:
Did I get it right that the key blocks can't be mixed with data blocks?

Reply:
Data blocks are placed as a leafes of that tree, it means we have to reach the lowest level of the tree. 
Iterating using BFS, if we reach first data blocks, it means all blocks on that level represents data.